### ISSUE TITLE
undeterministic behavior of capi

### ISSUE THREAD

**Issue by zhoonit** [nnstreamer/nntrainer#868]
from `ninja test`, occasionally (1 time out of 10?), I can observe
```bash
[ RUN      ] nntrainer_capi_nnmodel.train_01_p
../test/tizen_capi/unittest_tizen_capi.cpp:306: Failure
      Expected: status
      Which is: -22
To be equal to: ML_ERROR_NONE
      Which is: 0
../test/tizen_capi/unittest_tizen_capi.cpp:309: Failure
      Expected: status
      Which is: -22
To be equal to: ML_ERROR_NONE
      Which is: 0
../test/tizen_capi/unittest_tizen_capi.cpp:312: Failure
      Expected: status
      Which is: -22
To be equal to: ML_ERROR_NONE
      Which is: 0
../test/tizen_capi/unittest_tizen_capi.cpp:41: Failure
      Expected: status
      Which is: -22
To be equal to: ML_ERROR_NONE
      Which is: 0
-------
```

Sometimes, I get

```
[ RUN      ] nntrainer_ccapi.train_with_config_01_p
../test/ccapi/unittest_ccapi.cpp:166: Failure
      Expected: model->loadFromConfig(config_file)
      Which is: -22
To be equal to: ML_ERROR_NONE
      Which is: 0
../test/ccapi/unittest_ccapi.cpp:167: Failure
      Expected: model->compile()
      Which is: -22
To be equal to: ML_ERROR_NONE
      Which is: 0
../test/ccapi/unittest_ccapi.cpp:168: Failure
      Expected: model->initialize()
      Which is: -1073741822
To be equal to: ML_ERROR_NONE
      Which is: 0
../test/ccapi/unittest_ccapi.cpp:171: Failure
      Expected: model->getTrainingLoss()
      Which is: 0
To be equal to: 4.434051
../test/ccapi/unittest_ccapi.cpp:172: Failure
      Expected: model->getValidationLoss()
      Which is: 0
To be equal to: 2.9646113
../test/ccapi/unittest_ccapi.cpp:173: Failure
      Expected: model->getLoss()
      Which is: 0
To be equal to: 2.8036859
[  FAILED  ] nntrainer_ccapi.train_with_config_01_p (14 ms)
[ RUN      ] nntrainer_ccapi.train_dataset_with_file_01_p
#1/2 [ =                    ] 0% ( Training Loss: 2.38573 )
#1/2 [ =======              ] 31% ( Training Loss: 2.37778 )
#1/2 [ =============        ] 63% ( Training Loss: 2.29484 )
#1/2 - Training Loss: 2.35278 >> [ Accuracy: 6.25% - Validation Loss : 2.3133 ] 
#2/2 [ =                    ] 0% ( Training Loss: 2.09096 )
#2/2 [ =======              ] 31% ( Training Loss: 2.23402 )
#2/2 [ =============        ] 63% ( Training Loss: 2.21693 )
#2/2 - Training Loss: 2.18064 >> [ Accuracy: 16.6667% - Validation Loss : 2.25438 ] 
../test/ccapi/unittest_ccapi.cpp:221: Failure
      Expected: model->getTrainingLoss()
      Which is: 2.1806376
To be equal to: 2.1934659
../test/ccapi/unittest_ccapi.cpp:222: Failure
      Expected: model->getValidationLoss()
      Which is: 2.2543814
To be equal to: 2.2051108
../test/ccapi/unittest_ccapi.cpp:223: Failure
      Expected: model->getLoss()
      Which is: 2.2429025
To be equal to: 2.1751201
[  FAILED  ] nntrainer_ccapi.train_dataset_with_file_01_p (1283 ms)
[ RUN      ] nntrainer_ccapi.train_dataset_with_generator_01_p
#1/2 [          0           ]  ( Training Loss: 2.30068 )
#1/2 [          16          ]  ( Training Loss: 2.40239 )
#1/2 [          32          ]  ( Training Loss: 2.46805 )
#1/2 - Training Loss: 2.39037 >> [ Accuracy: 39.5833% - Validation Loss : 2.16501 ] 
#2/2 [          0           ]  ( Training Loss: 2.11128 )
#2/2 [          16          ]  ( Training Loss: 2.2695 )
#2/2 [          32          ]  ( Training Loss: 2.18525 )
#2/2 - Training Loss: 2.18867 >> [ Accuracy: 60.4167% - Validation Loss : 1.96841 ] 
../test/ccapi/unittest_ccapi.cpp:270: Failure
      Expected: model->getTrainingLoss()
      Which is: 2.1886747
To be equal to: 2.2109976
../test/ccapi/unittest_ccapi.cpp:271: Failure
      Expected: model->getValidationLoss()
      Which is: 1.9684099
To be equal to: 1.995334
../test/ccapi/unittest_ccapi.cpp:272: Failure
      Expected: model->getLoss()
      Which is: 1.9472635
To be equal to: 1.9843285
[  FAILED  ] nntrainer_ccapi.train_dataset_with_generator_01_p (1063 ms)
```

**Comment by kparichay** [comment]
Is this with tizen or ubuntu, or does this happen with both?

**Comment by zhoonit** [comment]
1. tested only on ubuntu with meson build && ninja test
2. only tizen api fails occasionally
3. will do the loop and report back

**Comment by zhoonit** [comment]
Could you try this? You are expected to observe it fails 100% after the first iteration,
```
build/test/tizen_capi/unittest_tizen_capi --gtest_repeat=5
```
```
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] nntrainer_capi_nnmodel.train_01_p
[  FAILED  ] nntrainer_capi_nnmodel.train_with_file_01_p
[  FAILED  ] nntrainer_capi_nnmodel.train_with_generator_01_p
```

**Comment by kparichay** [comment]
It shoud fail.
The unittests in tizen capi start with a random data based on a fixed seed always which makes NNTrainer deterministic in all its run. The function `getSeed()` function in `util_func.cpp` provides the same seed when called 5 times from two different processes.
However, with `test/tizen_capi/unittest_tizen_capi --gtest_repeat=5`, its 1 process running the unittest 5 times, and so when a uni test runs the 2nd time, the `getSeed()` can produce the same seed (which is intended behavior).

There are two workarounds - 
1. We fix the init values and load it from file. But this will have to all the unittests where the outputs must be matched and this list will keep increasing.
2. Run the unittests in different processes by invoking `test/tizen_capi/unittest_tizen_capi` one by one.

Same argument applies for ccapi.

**Comment by zhoonit** [comment]
Acked, let's convert this to an issue then a bug

**Comment by zhoonit** [comment]
Actually, let's close this for now and if we have time deal with this. I think it's hard to say it's a bug or sth.

**Comment by kparichay** [comment]
> Actually, let's close this for now and if we have time deal with this. I think it's hard to say it's a bug or sth.

This is not a bug for the framework. Just a limitation of the way the unittest has been designed.
